File: python/file_io.py
import json, os, re
from functions import parse_tag, remove_repeats

def copy_component(path):
    file = open(path, "r")
    output = file.read()
    file.close()
    return output

def listdir_recursive(directory):
    html_files = []
    static_files = []
    directories = []
    dirs = [directory]
    while len(dirs) > 0:
        current = dirs[0]
        for item in os.listdir(current):
            item = os.path.join(current, item)
            if os.path.isdir(item):
                dirs.append(item)
                directories.append(item)
            elif file_ext(item) == ".html":
                html_files.append(item)
            else:
                static_files.append(item)
        del dirs[0]
    return html_files, static_files, directories

def read_json(path):
    return json.loads(open(path, "r").read())

def file_ext(path):
    return os.path.splitext(path)[1]

def write_component(comp_obj, path):
    file = open(path, "w")
    file.write(comp_obj)
    file.close()

File: js/file-io.js
const fs = require("fs");
const path = require("path");
const { PROJECT_DIR } = require("./runtime-values");

function copyComponent(filePath) {
    return fs.readFileSync(filePath).toString();
}

function listDirRecursive(directory) {
    const pagesToRender = [];
    const staticFiles = [];
    const directories = [];
    const dirs = [directory];
    while (dirs.length > 0) {
        var current = dirs[0];
        for (item of fs.readdirSync(current)) {
            item = current + "/" + item;
            if (fs.statSync(item).isDirectory()) {
                dirs.push(item);
                directories.push(item);
            }
            else if (path.extname(item) == ".html") {
                pagesToRender.push(item);
            }
            else {
                staticFiles.push(item);
            }
        }
        dirs.splice(0, 1);
    }
    return { pagesToRender, staticFiles, directories };
}

function writeComponent(compObj, filePath) {
    fs.writeFileSync(PROJECT_DIR + "/" + filePath, compObj);
}

function readJson(filePath) {
    return require(PROJECT_DIR + "/" + filePath);
}

module.exports = { copyComponent, listDirRecursive, readJson, writeComponent };
